File: apps/telomi/server/research/python-tools/tools/github.py
"""Read-only GitHub discovery and Runtime-owned download API."""

from __future__ import annotations

import re
from collections.abc import Sequence
from datetime import date
from difflib import get_close_matches
from hashlib import sha256
from typing import Any, Literal, TypedDict

from research_runtime import search_source, workspace_path

__all__ = [
    "GitHubDiscovery",
    "GitHubRecord",
    "clone_repository",
    "discover_repositories",
    "download_file",
    "download_release",
    "get_issue",
    "get_repository",
    "search_code",
    "search_issues",
    "search_repositories",
    "search_topics",
]

MAX_RESULTS = 100
DISCOVERY_RESULT_LIMIT = 100
DISCOVERY_PAGE_SIZE = 20
DISCOVERY_CACHE_LIMIT = 5
DISCOVERY_LANES = ("topic_stars", "created_range", "active", "keywords")
DISCOVERY_RECORD_FIELDS = (
    "full_name", "url", "description", "stars", "forks", "language", "topics",
    "license", "created_at", "pushed_at", "archived", "discovery_lanes",
)
REPOSITORY = re.compile(
    r"^[A-Za-z0-9](?:[A-Za-z0-9-]{0,38})/[A-Za-z0-9._-]{1,100}$"
)
TOPIC = re.compile(r"^[A-Za-z0-9](?:[A-Za-z0-9-]{0,49})$")


class GitHubRecord(TypedDict, total=False):
    id: str
    title: str
    url: str
    content_path: str
    snippet: str
    repository: str
    owner: str
    path: str
    sha: str
    issue_number: int
    body: str
    comments: list[dict[str, Any]]
    stars: int
    forks: int
    language: str
    updated_at: str
    download_path: str
    material_cache_hit: bool
    full_name: str
    description: str
    topics: list[str]
    license: object
    created_at: str
    pushed_at: str
    archived: bool
    name: str
    display_name: str
    short_description: str
    featured: bool
    curated: bool
    created_by: str
    discovery_lanes: list[dict[str, Any]]
    metadata: dict[str, Any]


class GitHubDiscovery(TypedDict):
    records: list[GitHubRecord]
    lane_counts: dict[str, int]
    unique_count: int
    returned_count: int
    next_cursor: str | None


_DISCOVERY_POOLS: dict[
    tuple[tuple[str, ...], str | None, str | None, str | None, str | None, int | None],
    tuple[list[GitHubRecord], dict[str, int]],
] = {}


def search_topics(
    query: str,
    *,
    curated_only: bool = False,
    max_results: int = 20,
) -> list[GitHubRecord]:
    """Search GitHub's topic catalog.

    Args:
        query: Topic name or phrase.
        curated_only: Restrict results to the small set GitHub itself curates
            (generic names such as deep-learning). Most domain topics such as
            text-to-speech are community topics, so leave this False.
        max_results: Maximum results, from 1 through 100.

    Returns:
        Topic records with names, descriptions, and curation metadata.

    Raises:
        ValueError: If an argument is invalid.
        ResearchRuntimeError: If Runtime or GitHub fails.
    """
    value = _text(query, "query", 2_000)
    if not isinstance(curated_only, bool):
        raise ValueError("curated_only must be a boolean")
    size = _limit(max_results)
    return _run(
        "search_topics",
        {"query": value, "curated_only": curated_only, "limit": size},
        value,
        size,
    )


def search_repositories(
    query: str,
    *,
    topics: Sequence[str] = (),
    language: str | None = None,
    min_stars: int | None = None,
    created_after: str | None = None,
    created_before: str | None = None,
    pushed_after: str | None = None,
    sort: Literal["stars", "updated", "forks"] = "stars",
    max_results: int = 20,
) -> list[GitHubRecord]:
    """Search public GitHub repositories.

    GitHub ANDs whitespace-separated query terms. Topic filters are the
    reliable discovery path; keep free-text queries short and supplemental.

    Args:
        query: GitHub repository search phrase. May be empty when topics are supplied.
        topics: GitHub topic names, all applied as native filters.
        language: Optional GitHub language filter.
        min_stars: Optional inclusive minimum star count.
        created_after: Optional inclusive creation-date lower bound.
        created_before: Optional inclusive creation-date upper bound.
        pushed_after: Optional inclusive last-push lower bound.
        sort: GitHub result sort: stars, updated, or forks.
        max_results: Maximum results, from 1 through 100.

    Returns:
        Repository candidate dictionaries.

    Raises:
        ValueError: If the query or result limit is invalid.
        ResearchRuntimeError: If Runtime or GitHub fails.
    """
    topic_values = _topics(topics)
    value = _query(query, topic_values)
    size = _limit(max_results)
    if sort not in ("stars", "updated", "forks"):
        raise ValueError("sort must be stars, updated, or forks")
    parameters: dict[str, Any] = {
        "query": value,
        "topics": topic_values,
        "sort": sort,
        "limit": size,
    }
    if language is not None:
        parameters["language"] = _text(language, "language", 100)
    if min_stars is not None:
        parameters["min_stars"] = _nonnegative_integer(min_stars, "min_stars")
    for name, bound in (
        ("created_after", created_after),
        ("created_before", created_before),
        ("pushed_after", pushed_after),
    ):
        if bound is not None:
            parameters[name] = _date(bound, name)
    if created_after and created_before and parameters["created_after"] > parameters["created_before"]:
        raise ValueError("created_after must not be later than created_before")
    return _run("search_repositories", parameters, value or " ".join(topic_values), size)


def discover_repositories(
    topics: str | Sequence[str],
    *,
    keywords: str | None = None,
    start_date: str | None = None,
    end_date: str | None = None,
    language: str | None = None,
    min_stars: int | None = None,
    cursor: str | None = None,
    purpose: str = "Discover GitHub repositories through fixed native lanes",
) -> GitHubDiscovery:
    """Build one bounded repository pool from exact GitHub topics and fixed lanes.

    Args:
        topics: One or more exact GitHub topic names, as returned by search_topics().
        keywords: Optional short supplemental query.
        start_date: Optional inclusive creation and activity lower bound.
        end_date: Optional inclusive creation upper bound.
        language: Optional GitHub language filter applied to every lane.
        min_stars: Optional inclusive star threshold applied to every lane.
        cursor: Exact ``next_cursor`` from the preceding discovery page.
        purpose: Short provenance note.

    Returns:
        One compact record page, lane counts, pool counts, and next cursor.

    Raises:
        ValueError: If topics, dates, filters, or cursor are invalid.
        ResearchRuntimeError: If Runtime or GitHub fails.
    """
    topic_values = _topics([topics] if isinstance(topics, str) else topics)
    if not topic_values:
        raise ValueError("topics must contain at least one topic name")
    if (start_date is None) != (end_date is None):
        raise ValueError("start_date and end_date must be provided together")
    start = _date(start_date, "start_date") if start_date is not None else None
    end = _date(end_date, "end_date") if end_date is not None else None
    if start and end and start > end:
        raise ValueError("start_date must not be later than end_date")
    keyword_value = _text(keywords, "keywords", 2_000) if keywords is not None else None
    language_value = _text(language, "language", 100) if language is not None else None
    stars_value = _nonnegative_integer(min_stars, "min_stars") if min_stars is not None else None
    cache_key = (tuple(topic_values), keyword_value, start, end, language_value, stars_value)
    key_hash = sha256(repr(cache_key).encode()).hexdigest()[:16]
    offset = _discovery_offset(cursor, key_hash)
    cached = _DISCOVERY_POOLS.get(cache_key)
    if cached is not None:
        return _discovery_page(*cached, offset, key_hash)
    if cursor is not None:
        raise ValueError("cursor discovery pool is no longer cached; restart without a cursor")

    for topic in topic_values:
        matches = search_topics(topic, curated_only=False, max_results=20)
        names = [row["name"] for row in matches if isinstance(row.get("name"), str)]
        if topic not in names:
            closest = get_close_matches(topic, names, n=5, cutoff=0.3) or names[:5]
            suffix = f" Closest GitHub topics: {', '.join(closest)}." if closest else ""
            raise ValueError(f"Unknown GitHub topic '{topic}'.{suffix}")

    records_by_name: dict[str, GitHubRecord] = {}
    lane_counts: dict[str, int] = {}
    lane_ids = {lane: {topic: [] for topic in topic_values} for lane in DISCOVERY_LANES}
    common = {"language": language_value, "min_stars": stars_value, "max_results": 100}
    for topic in topic_values:
        lanes: list[tuple[str, list[GitHubRecord]]] = [
            ("topic_stars", search_repositories("", topics=[topic], sort="stars", **common)),
        ]
        if start and end:
            lanes.extend([
                ("created_range", search_repositories(
                    "", topics=[topic], created_after=start, created_before=end, sort="stars", **common,
                )),
                ("active", search_repositories(
                    "", topics=[topic], pushed_after=start, sort="updated", **common,
                )),
            ])
        if keyword_value:
            lanes.append(("keywords", search_repositories(
                keyword_value, topics=[topic], sort="stars", **common,
            )))
        for lane, rows in lanes:
            lane_counts[f"{topic}:{lane}"] = len(rows)
            for rank, row in enumerate(rows, start=1):
                record = _compact_repository(row)
                full_name = record.get("full_name")
                if not isinstance(full_name, str) or not full_name:
                    continue
                stored = records_by_name.setdefault(full_name, {**record, "discovery_lanes": []})
                stored["discovery_lanes"].append({"topic": topic, "lane": lane, "rank": rank})
                lane_ids[lane][topic].append(full_name)

    selected_names: list[str] = []
    seen: set[str] = set()
    for lane in DISCOVERY_LANES:
        rows_by_topic = lane_ids[lane]
        ordered = [
            rows[rank]
            for rank in range(max((len(rows) for rows in rows_by_topic.values()), default=0))
            for rows in rows_by_topic.values()
            if rank < len(rows)
        ]
        for full_name in ordered:
            if len(selected_names) == DISCOVERY_RESULT_LIMIT:
                break
            if full_name not in seen:
                seen.add(full_name)
                selected_names.append(full_name)
    selected = [records_by_name[full_name] for full_name in selected_names]
    if len(_DISCOVERY_POOLS) >= DISCOVERY_CACHE_LIMIT:
        _DISCOVERY_POOLS.pop(next(iter(_DISCOVERY_POOLS)))
    _DISCOVERY_POOLS[cache_key] = (selected, lane_counts)
    return _discovery_page(selected, lane_counts, offset, key_hash)


def get_repository(repository: str) -> list[GitHubRecord]:
    """Fetch one repository and its stable metadata.

    Args:
        repository: Repository in ``OWNER/REPO`` form.

    Returns:
        A single repository record when found.

    Raises:
        ValueError: If the repository name is invalid.
        ResearchRuntimeError: If Runtime or GitHub fails.
    """
    value = _repository(repository)
    return _run("get_repository", {"repository": value}, f"repository:{value}", 1)


def search_code(
    query: str,
    *,
    repository: str | None = None,
    max_results: int = 20,
) -> list[GitHubRecord]:
    """Search code, optionally within one repository.

    Args:
        query: GitHub code search phrase or qualifiers.
        repository: Optional repository in ``OWNER/REPO`` form.
        max_results: Maximum results, from 1 through 100.

    Returns:
        Matching code records with repository, path, SHA, and text fragments.

    Raises:
        ValueError: If an argument is invalid.
        ResearchRuntimeError: If Runtime or GitHub fails.
    """
    value = _text(query, "query", 2_000)
    size = _limit(max_results)
    parameters: dict[str, Any] = {"query": value, "limit": size}
    if repository is not None:
        parameters["repository"] = _repository(repository)
    return _run("search_code", parameters, value, size)


def search_issues(
    query: str,
    *,
    repository: str | None = None,
    state: Literal["open", "closed", "all"] = "all",
    match: Literal["title", "body", "comments"] | None = None,
    max_results: int = 20,
) -> list[GitHubRecord]:
    """Search GitHub Issues without loading every discussion.

    Args:
        query: Issue search phrase or qualifiers.
        repository: Optional repository in ``OWNER/REPO`` form.
        state: ``open``, ``closed``, or ``all``.
        match: Restrict matching to issue title, body, or comments.
        max_results: Maximum results, from 1 through 100.

    Returns:
        Issue candidates. Call :func:`get_issue` for full comments.

    Raises:
        ValueError: If an argument is invalid.
        ResearchRuntimeError: If Runtime or GitHub fails.
    """
    value = _text(query, "query", 2_000)
    size = _limit(max_results)
    if state not in ("open", "closed", "all"):
        raise ValueError("state must be open, closed, or all")
    if match is not None and match not in ("title", "body", "comments"):
        raise ValueError("match must be title, body, or comments")
    parameters: dict[str, Any] = {
        "query": value,
        "state": state,
        "limit": size,
    }
    if repository is not None:
        parameters["repository"] = _repository(repository)
    if match is not None:
        parameters["match"] = match
    return _run("search_issues", parameters, value, size)


def get_issue(repository: str, number: int) -> list[GitHubRecord]:
    """Fetch one Issue body and its complete comment thread.

    Args:
        repository: Repository in ``OWNER/REPO`` form.
        number: Positive Issue number.

    Returns:
        One Issue record containing structured ``comments``.

    Raises:
        ValueError: If the repository or Issue number is invalid.
        ResearchRuntimeError: If Runtime or GitHub fails.
    """
    value = _repository(repository)
    if not isinstance(number, int) or isinstance(number, bool) or number <= 0:
        raise ValueError("number must be a positive integer")
    return _run(
        "get_issue",
        {"repository": value, "number": number},
        f"issue:{value}#{number}",
        1,
    )


def clone_repository(
    repository: str,
    *,
    ref: str | None = None,
    full_history: bool = False,
) -> list[GitHubRecord]:
    """Clone a repository into the current PrimeSearch workspace.

    Args:
        repository: Repository in ``OWNER/REPO`` form.
        ref: Optional branch, tag, or commit reference.
        full_history: Clone all history instead of the default depth-one clone.

    Returns:
        One record whose ``download_path`` is a Runtime-owned directory.

    Raises:
        ValueError: If an argument is invalid.
        ResearchRuntimeError: If Runtime, GitHub CLI, or Git fails.
    """
    value = _repository(repository)
    if not isinstance(full_history, bool):
        raise ValueError("full_history must be a boolean")
    parameters: dict[str, Any] = {
        "repository": value,
        "full_history": full_history,
    }
    if ref is not None:
        parameters["ref"] = _ref(ref, "ref")
    return _run("clone_repository", parameters, f"clone:{value}:{ref or 'default'}", 1)


def download_release(
    repository: str,
    *,
    tag: str | None = None,
    patterns: Sequence[str] = (),
    archive: Literal["zip", "tar.gz"] | None = None,
) -> list[GitHubRecord]:
    """Download release assets or a source archive into the workspace.

    Args:
        repository: Repository in ``OWNER/REPO`` form.
        tag: Optional release tag. Without one, patterns or archive is required.
        patterns: Optional asset glob patterns.
        archive: Optional source archive format, ``zip`` or ``tar.gz``.

    Returns:
        One record for each downloaded file with ``download_path``.

    Raises:
        ValueError: If an argument is invalid.
        ResearchRuntimeError: If Runtime or GitHub CLI fails.
    """
    value = _repository(repository)
    normalized_patterns = _patterns(patterns)
    normalized_tag = _ref(tag, "tag") if tag is not None else None
    if archive is not None and archive not in ("zip", "tar.gz"):
        raise ValueError("archive must be zip or tar.gz")
    if normalized_tag is None and not normalized_patterns and archive is None:
        raise ValueError("tag, patterns, or archive is required")
    parameters: dict[str, Any] = {"repository": value}
    if normalized_tag is not None:
        parameters["tag"] = normalized_tag
    if normalized_patterns:
        parameters["patterns"] = normalized_patterns
    if archive is not None:
        parameters["archive"] = archive
    return _run(
        "download_release",
        parameters,
        f"release:{value}:{normalized_tag or 'latest'}",
        MAX_RESULTS,
    )


def download_file(
    repository: str,
    path: str,
    *,
    ref: str | None = None,
) -> list[GitHubRecord]:
    """Download one repository file into the workspace.

    Args:
        repository: Repository in ``OWNER/REPO`` form.
        path: Safe relative file path inside the repository.
        ref: Optional branch, tag, or commit reference.

    Returns:
        One file record with ``download_path``.

    Raises:
        ValueError: If an argument is invalid.
        ResearchRuntimeError: If Runtime or GitHub CLI fails.
    """
    value = _repository(repository)
    relative_path = _path(path)
    parameters: dict[str, Any] = {
        "repository": value,
        "path": relative_path,
    }
    if ref is not None:
        parameters["ref"] = _ref(ref, "ref")
    return _run(
        "download_file",
        parameters,
        f"file:{value}:{relative_path}:{ref or 'default'}",
        1,
    )


def _run(
    operation: str,
    parameters: dict[str, Any],
    query: str,
    max_results: int,
) -> list[GitHubRecord]:
    rows = search_source([{
        "query": query,
        "purpose": f"Host-owned GitHub {operation}",
        "max_results": max_results,
        "provider_request": {
            "operation": operation,
            "parameters": parameters,
        },
    }], source="github")
    return [_enrich(row) for row in rows]


def _enrich(row: dict[str, Any]) -> GitHubRecord:
    output = dict(row)
    metadata = row.get("metadata")
    if not isinstance(metadata, dict):
        return output  # type: ignore[return-value]
    for key in (
        "repository",
        "owner",
        "path",
        "sha",
        "issue_number",
        "body",
        "comments",
        "stars",
        "forks",
        "language",
        "updated_at",
        "topics",
        "license",
        "created_at",
        "pushed_at",
        "archived",
        "name",
        "display_name",
        "short_description",
        "featured",
        "curated",
        "created_by",
        "material_cache_hit",
    ):
        if key in metadata:
            output[key] = metadata[key]
    artifact_path = metadata.get("artifact_path")
    if isinstance(artifact_path, str) and artifact_path:
        output["download_path"] = workspace_path(artifact_path)
    return output  # type: ignore[return-value]


def _compact_repository(row: GitHubRecord) -> GitHubRecord:
    full_name = row.get("full_name") or row.get("repository") or row.get("title")
    values: dict[str, Any] = {
        "full_name": full_name,
        "url": row.get("url"),
        "description": row.get("description", row.get("snippet", "")),
        **{key: row[key] for key in DISCOVERY_RECORD_FIELDS if key in row},
    }
    return {key: values[key] for key in DISCOVERY_RECORD_FIELDS if key in values}  # type: ignore[return-value]


def _discovery_page(
    selected: list[GitHubRecord],
    lane_counts: dict[str, int],
    offset: int,
    key_hash: str,
) -> GitHubDiscovery:
    if offset and offset >= len(selected):
        raise ValueError("cursor is beyond the current discovery pool; restart without a cursor")
    end = min(offset + DISCOVERY_PAGE_SIZE, len(selected))
    records = [record.copy() for record in selected[offset:end]]
    return {
        "records": records,
        "lane_counts": lane_counts,
        "unique_count": len(selected),
        "returned_count": len(records),
        "next_cursor": f"github-discovery:{key_hash}:{end}" if end < len(selected) else None,
    }


def _discovery_offset(cursor: str | None, key_hash: str) -> int:
    if cursor is None:
        return 0
    value = _text(cursor, "cursor", 64)
    parts = value.split(":")
    if len(parts) != 3 or parts[0] != "github-discovery" or not parts[2].isdigit() or int(parts[2]) <= 0:
        raise ValueError("cursor must be the exact next_cursor returned by discover_repositories()")
    if parts[1] != key_hash:
        raise ValueError("cursor was created for different discovery parameters")
    return int(parts[2])


def _repository(value: str) -> str:
    result = _text(value, "repository", 140)
    if not REPOSITORY.fullmatch(result):
        raise ValueError("repository must use OWNER/REPO form")
    return result


def _path(value: str) -> str:
    result = _text(value, "path", 4_096)
    parts = result.split("/")
    if result.startswith("/") or any(part in ("", ".", "..") for part in parts):
        raise ValueError("path must be a safe relative repository path")
    return result


def _ref(value: str, name: str) -> str:
    result = _text(value, name, 255)
    if result.startswith("-") or any(ord(character) < 32 or ord(character) == 127 for character in result):
        raise ValueError(f"{name} is invalid")
    return result


def _patterns(values: Sequence[str]) -> list[str]:
    if isinstance(values, (str, bytes)) or len(values) > 20:
        raise ValueError("patterns must be a sequence of at most 20 strings")
    return list(dict.fromkeys(_text(value, "pattern", 256) for value in values))


def _topics(values: Sequence[str]) -> list[str]:
    if isinstance(values, (str, bytes)) or len(values) > 20:
        raise ValueError("topics must be a sequence of at most 20 topic names")
    topics = list(dict.fromkeys(_text(value, "topic", 50) for value in values))
    if any(not TOPIC.fullmatch(topic) for topic in topics):
        raise ValueError("topic names may contain only letters, numbers, and hyphens")
    return topics


def _query(value: str, topics: Sequence[str]) -> str:
    if not isinstance(value, str) or len(value) > 2_000:
        raise ValueError("query must be a string up to 2000 characters")
    result = value.strip()
    if not result and not topics:
        raise ValueError("query may be empty only when topics are supplied")
    return result


def _date(value: str, name: str) -> str:
    result = _text(value, name, 10)
    try:
        parsed = date.fromisoformat(result)
    except ValueError:
        raise ValueError(f"{name} must use YYYY-MM-DD") from None
    if parsed.isoformat() != result:
        raise ValueError(f"{name} must use YYYY-MM-DD")
    return result


def _nonnegative_integer(value: int, name: str) -> int:
    if not isinstance(value, int) or isinstance(value, bool) or value < 0:
        raise ValueError(f"{name} must be a non-negative integer")
    return value


def _limit(value: int) -> int:
    if not isinstance(value, int) or isinstance(value, bool) or not 1 <= value <= MAX_RESULTS:
        raise ValueError(f"max_results must be between 1 and {MAX_RESULTS}")
    return value


def _text(value: str, name: str, max_length: int) -> str:
    if not isinstance(value, str) or not value.strip() or len(value) > max_length:
        raise ValueError(
            f"{name} must be a non-empty string up to {max_length} characters"
        )
    return value.strip()

File: apps/telomi/server/research/sources/contracts/github.ts
import type { ResearchProviderRequest } from "../../../providers/search-types.js";

const MAX_RESULTS = 100;
const REPOSITORY = /^[A-Za-z0-9](?:[A-Za-z0-9-]{0,38})\/[A-Za-z0-9._-]{1,100}$/u;
const TOPIC = /^[A-Za-z0-9](?:[A-Za-z0-9-]{0,49})$/u;

export const GITHUB_OPERATIONS = [
	"search_topics",
	"search_repositories",
	"get_repository",
	"search_code",
	"search_issues",
	"get_issue",
	"clone_repository",
	"download_release",
	"download_file",
] as const;

export type GitHubOperation = typeof GITHUB_OPERATIONS[number];

export interface GitHubParametersByOperation {
	search_topics: { query: string; curated_only: boolean; limit: number };
	search_repositories: {
		query: string;
		topics?: string[];
		language?: string;
		min_stars?: number;
		created_after?: string;
		created_before?: string;
		pushed_after?: string;
		sort?: "stars" | "updated" | "forks";
		order?: "desc" | "asc";
		limit: number;
	};
	get_repository: { repository: string };
	search_code: { query: string; repository?: string; limit: number };
	search_issues: {
		query: string;
		repository?: string;
		state: "open" | "closed" | "all";
		match?: "title" | "body" | "comments";
		limit: number;
	};
	get_issue: { repository: string; number: number };
	clone_repository: { repository: string; ref?: string; full_history: boolean };
	download_release: {
		repository: string;
		tag?: string;
		patterns?: string[];
		archive?: "zip" | "tar.gz";
	};
	download_file: { repository: string; path: string; ref?: string };
}

export type ParsedGitHubProviderRequest = {
	[K in GitHubOperation]: {
		operation: K;
		parameters: GitHubParametersByOperation[K];
	};
}[GitHubOperation];

export function parseGitHubProviderRequest(
	request: ResearchProviderRequest | undefined,
	requestQuery: string,
	requestMaxResults: number,
): ParsedGitHubProviderRequest {
	if (!request) throw new Error("GitHub Provider requests require an explicit operation");
	if (!GITHUB_OPERATIONS.includes(request.operation as GitHubOperation)) {
		throw new Error(`Unsupported GitHub operation '${request.operation}'`);
	}
	const operation = request.operation as GitHubOperation;
	const row = strictObject(request.parameters, allowedKeys(operation), operation);
	const defaultLimit = Math.min(integer(requestMaxResults, "max_results", 1, MAX_RESULTS), MAX_RESULTS);

	if (operation === "search_topics") {
		return {
			operation,
			parameters: {
				query: optionalText(row.query, "query", 2_000) ?? text(requestQuery, "query", 2_000),
				curated_only: optionalBoolean(row.curated_only, "curated_only") ?? false,
				limit: optionalInteger(row.limit, "limit", 1, MAX_RESULTS) ?? defaultLimit,
			},
		};
	}
	if (operation === "search_repositories") {
		const topics = optionalTextArray(row.topics, "topics", 20, 50)?.map(topic);
		const query = row.query === undefined
			? text(requestQuery, "query", 2_000)
			: possiblyEmptyText(row.query, "query", 2_000);
		if (!query && !topics?.length) throw new Error("GitHub query may be empty only when topics are provided");
		const createdAfter = row.created_after === undefined ? undefined : dateText(row.created_after, "created_after");
		const createdBefore = row.created_before === undefined ? undefined : dateText(row.created_before, "created_before");
		if (createdAfter && createdBefore && createdAfter > createdBefore) {
			throw new Error("GitHub created_after must not be later than created_before");
		}
		return {
			operation,
			parameters: {
				query,
				...(topics?.length ? { topics } : {}),
				...(row.language === undefined ? {} : { language: text(row.language, "language", 100) }),
				...(row.min_stars === undefined ? {} : {
					min_stars: integer(row.min_stars, "min_stars", 0, Number.MAX_SAFE_INTEGER),
				}),
				...(createdAfter ? { created_after: createdAfter } : {}),
				...(createdBefore ? { created_before: createdBefore } : {}),
				...(row.pushed_after === undefined ? {} : {
					pushed_after: dateText(row.pushed_after, "pushed_after"),
				}),
				...(row.sort === undefined ? {} : {
					sort: requiredEnum(row.sort, "sort", ["stars", "updated", "forks"] as const),
				}),
				...(row.order === undefined ? {} : {
					order: requiredEnum(row.order, "order", ["desc", "asc"] as const),
				}),
				limit: optionalInteger(row.limit, "limit", 1, MAX_RESULTS) ?? defaultLimit,
			},
		};
	}
	if (operation === "get_repository") {
		return { operation, parameters: { repository: repository(row.repository) } };
	}
	if (operation === "search_code") {
		return {
			operation,
			parameters: {
				query: optionalText(row.query, "query", 2_000) ?? text(requestQuery, "query", 2_000),
				...(row.repository === undefined ? {} : { repository: repository(row.repository) }),
				limit: optionalInteger(row.limit, "limit", 1, MAX_RESULTS) ?? defaultLimit,
			},
		};
	}
	if (operation === "search_issues") {
		return {
			operation,
			parameters: {
				query: optionalText(row.query, "query", 2_000) ?? text(requestQuery, "query", 2_000),
				...(row.repository === undefined ? {} : { repository: repository(row.repository) }),
				state: optionalEnum(row.state, "state", ["open", "closed", "all"] as const) ?? "all",
				...(row.match === undefined ? {} : {
					match: requiredEnum(row.match, "match", ["title", "body", "comments"] as const),
				}),
				limit: optionalInteger(row.limit, "limit", 1, MAX_RESULTS) ?? defaultLimit,
			},
		};
	}
	if (operation === "get_issue") {
		return {
			operation,
			parameters: {
				repository: repository(row.repository),
				number: integer(row.number, "number", 1, Number.MAX_SAFE_INTEGER),
			},
		};
	}
	if (operation === "clone_repository") {
		return {
			operation,
			parameters: {
				repository: repository(row.repository),
				...(row.ref === undefined ? {} : { ref: ref(row.ref, "ref") }),
				full_history: optionalBoolean(row.full_history, "full_history") ?? false,
			},
		};
	}
	if (operation === "download_release") {
		const tag = row.tag === undefined ? undefined : ref(row.tag, "tag");
		const patterns = optionalTextArray(row.patterns, "patterns", 20, 256);
		const archive = row.archive === undefined
			? undefined
			: requiredEnum(row.archive, "archive", ["zip", "tar.gz"] as const);
		if (!tag && !patterns?.length && !archive) {
			throw new Error("GitHub download_release requires tag, patterns, or archive");
		}
		return {
			operation,
			parameters: {
				repository: repository(row.repository),
				...(tag ? { tag } : {}),
				...(patterns?.length ? { patterns } : {}),
				...(archive ? { archive } : {}),
			},
		};
	}
	return {
		operation,
		parameters: {
			repository: repository(row.repository),
			path: repositoryPath(row.path),
			...(row.ref === undefined ? {} : { ref: ref(row.ref, "ref") }),
		},
	};
}

export function canonicalGitHubQuery(request: ParsedGitHubProviderRequest): string {
	return `github:${request.operation}:${JSON.stringify(request.parameters)}`;
}

export function githubOperationMaterializesWorkspace(operation: string | undefined): boolean {
	return operation === "clone_repository" || operation === "download_release" || operation === "download_file";
}

function allowedKeys(operation: GitHubOperation): string[] {
	switch (operation) {
		case "search_topics": return ["query", "curated_only", "limit"];
		case "search_repositories": return [
			"query", "topics", "language", "min_stars", "created_after", "created_before",
			"pushed_after", "sort", "order", "limit",
		];
		case "get_repository": return ["repository"];
		case "search_code": return ["query", "repository", "limit"];
		case "search_issues": return ["query", "repository", "state", "match", "limit"];
		case "get_issue": return ["repository", "number"];
		case "clone_repository": return ["repository", "ref", "full_history"];
		case "download_release": return ["repository", "tag", "patterns", "archive"];
		case "download_file": return ["repository", "path", "ref"];
	}
}

function strictObject(
	value: unknown,
	allowedKeys: string[],
	operation: GitHubOperation,
): Record<string, unknown> {
	if (!value || typeof value !== "object" || Array.isArray(value)) {
		throw new Error("GitHub parameters must be an object");
	}
	const row = value as Record<string, unknown>;
	const allowed = new Set(allowedKeys);
	for (const key of Object.keys(row)) {
		if (!allowed.has(key)) throw new Error(`Unsupported GitHub ${operation} parameter '${key}'`);
	}
	return row;
}

function repository(value: unknown): string {
	const result = text(value, "repository", 140);
	if (!REPOSITORY.test(result)) throw new Error("GitHub repository must use OWNER/REPO form");
	return result;
}

function repositoryPath(value: unknown): string {
	const result = text(value, "path", 4_096);
	if (result.startsWith("/") || result.split("/").some((part) => !part || part === "." || part === "..")) {
		throw new Error("GitHub path must be a safe relative repository path");
	}
	return result;
}

function ref(value: unknown, name: string): string {
	const result = text(value, name, 255);
	if (result.startsWith("-") || /[\u0000-\u001f\u007f]/u.test(result)) {
		throw new Error(`GitHub ${name} is invalid`);
	}
	return result;
}

function text(value: unknown, name: string, maxLength: number): string {
	if (typeof value !== "string" || !value.trim() || value.length > maxLength) {
		throw new Error(`GitHub ${name} must be a non-empty string up to ${maxLength} characters`);
	}
	return value.trim();
}

function optionalText(value: unknown, name: string, maxLength: number): string | undefined {
	return value === undefined ? undefined : text(value, name, maxLength);
}

function possiblyEmptyText(value: unknown, name: string, maxLength: number): string {
	if (typeof value !== "string" || value.length > maxLength) {
		throw new Error(`GitHub ${name} must be a string up to ${maxLength} characters`);
	}
	return value.trim();
}

function topic(value: string): string {
	if (!TOPIC.test(value)) throw new Error("GitHub topic must contain only letters, numbers, and hyphens");
	return value;
}

function dateText(value: unknown, name: string): string {
	const result = text(value, name, 10);
	const parsed = new Date(`${result}T00:00:00Z`);
	if (!/^\d{4}-\d{2}-\d{2}$/u.test(result) || Number.isNaN(parsed.valueOf())
		|| parsed.toISOString().slice(0, 10) !== result) {
		throw new Error(`GitHub ${name} must use YYYY-MM-DD`);
	}
	return result;
}

function integer(value: unknown, name: string, minimum: number, maximum: number): number {
	if (!Number.isInteger(value) || Number(value) < minimum || Number(value) > maximum) {
		throw new Error(`GitHub ${name} must be an integer between ${minimum} and ${maximum}`);
	}
	return Number(value);
}

function optionalInteger(
	value: unknown,
	name: string,
	minimum: number,
	maximum: number,
): number | undefined {
	return value === undefined ? undefined : integer(value, name, minimum, maximum);
}

function optionalBoolean(value: unknown, name: string): boolean | undefined {
	if (value === undefined) return undefined;
	if (typeof value !== "boolean") throw new Error(`GitHub ${name} must be a boolean`);
	return value;
}

function requiredEnum<const T extends readonly string[]>(
	value: unknown,
	name: string,
	values: T,
): T[number] {
	if (typeof value !== "string" || !values.includes(value)) {
		throw new Error(`GitHub ${name} must be one of ${values.join(", ")}`);
	}
	return value as T[number];
}

function optionalEnum<const T extends readonly string[]>(
	value: unknown,
	name: string,
	values: T,
): T[number] | undefined {
	return value === undefined ? undefined : requiredEnum(value, name, values);
}

function optionalTextArray(
	value: unknown,
	name: string,
	maxItems: number,
	maxLength: number,
): string[] | undefined {
	if (value === undefined) return undefined;
	if (!Array.isArray(value) || value.length > maxItems) {
		throw new Error(`GitHub ${name} must be an array with at most ${maxItems} items`);
	}
	return [...new Set(value.map((item) => text(item, name, maxLength)))];
}
